pkg/api/unversioned/group_version.go

```diff
@@ -0,0 +1,75 @@
+/*
+Copyright 2015 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package unversioned
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+)
+
+// TODO: We need to remove the GroupVersion in types.go. We use the name GroupAndVersion here temporarily.
+type GroupAndVersion struct {
+	Group   string
+	Version string
+}
+
+func (gv *GroupAndVersion) String() string {
+	// special case of "v1" for backward compatibility
+	if gv.Group == "" {
+		return gv.Version
+	} else {
+		return gv.Group + "/" + gv.Version
+	}
+}
+
+func ParseGroupVersion(gv string) (GroupAndVersion, error) {
+	s := strings.Split(gv, "/")
+	// "v1" is the only special case. Otherwise GroupVersion is expected to contain
+	// one "/" dividing the string into two parts.
+	if len(s) == 1 && gv == "v1" {
+		return GroupAndVersion{"", "v1"}, nil
+	} else if len(s) == 2 {
+		return GroupAndVersion{s[0], s[1]}, nil
+	} else {
+		return GroupAndVersion{}, fmt.Errorf("Unexpected GroupVersion string: %v", gv)
+	}
+}
+
+// MarshalJSON implements the json.Marshaller interface.
+func (gv GroupAndVersion) MarshalJSON() ([]byte, error) {
+	s := gv.String()
+	if strings.Count(s, "/") > 1 {
```

Comment:
The condition is true, but why would gv.String() produce something that
violates the condition? I mean, it also doesn't hurt to check it, I guess.

On Thu, Nov 5, 2015 at 10:42 AM, Chao Xu <notifications@github.com> wrote:

> In pkg/api/unversioned/group_version.go
> <https://github.com/kubernetes/kubernetes/pull/16834#discussion_r44050629>
> :
>
> > +	s := strings.Split(gv, "/")
> > +	// "v1" is the only special case. Otherwise GroupVersion is expected to contain
> > +	// one "/" dividing the string into two parts.
> > +	if len(s) == 1 && gv == "v1" {
> > +		return GroupAndVersion{"", "v1"}, nil
> > +	} else if len(s) == 2 {
> > +		return GroupAndVersion{s[0], s[1]}, nil
> > +	} else {
> > +		return GroupAndVersion{}, fmt.Errorf("Unexpected GroupVersion string: %v", gv)
> > +	}
> > +}
> > +
> > +// MarshalJSON implements the json.Marshaller interface.
> > +func (gv GroupAndVersion) MarshalJSON() ([]byte, error) {
> > +	s := gv.String()
> > +	if strings.Count(s, "/") > 1 {
>
> Do you mean this condition does not hold anymore or I should move this
> check to somewhere else?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/16834/files#r44050629>.
>